### PR TITLE
distance to regular expression

### PR DESCRIPTION
**Pull request by jgaleotti**
Implementation of TT for Java Regular Expressions

### CODE REVIEW

File: client-java/instrumentation/pom.xml [hunk 1]
```diff
@@ -47,6 +47,16 @@
...
+        <!-- Dependencies for Java Regex distances -->
+        <dependency>
+            <groupId>net.sf.jgrapht</groupId>
+            <artifactId>jgrapht</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>dk.brics.automaton</groupId>
+            <artifactId>automaton</artifactId>
+        </dependency>
+
```
Comment (arcuri82) [inline client-java/instrumentation/pom.xml]: these libraries should be added as well to the shade plugin

File: core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt [hunk 2]
```diff
@@ -44,9 +43,14 @@ class CharacterClassEscapeRxGene(
     }
 
     override fun standardMutation(randomness: Randomness, apc: AdaptiveParameterControl, allGenes: List<Gene>) {
+        if (value=="") {
+            // if standardMutation was invoked before calling to randomize
+            // then do a randomize first
+            randomize(randomness, forceNewValue = true, allGenes = allGenes)
+        }
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt]: i think this should be removed. if it does happen that standard mutation is called before a randomized, then it is a bug. we should always randomize before an individual can be mutated. so we should rather investigate why it was not the case (ie bug)

File: core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt [hunk 3]
```diff
@@ -44,9 +43,14 @@ class CharacterClassEscapeRxGene(
...
 
         value = when(type){
-            "d" -> ((value.toInt() + randomness.choose(listOf(1,-1) + 10)) % 10).toString()
+            "d" -> ((Math.abs(value.toInt() + randomness.choose(listOf(1,-1) + 10))) % 10).toString()
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt]: why was `abs` necessary? can it ever happen that it is negative?

File: client-java/instrumentation/pom.xml (hunk 1 above)
Comment (jgaleotti) [reply client-java/instrumentation/pom.xml]: ok

File: core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt (hunk 3 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt]: yes. I found cases where the value is set to 0 and the randomness chose -1.
The result of -1 % 10 is -1.

Comment (arcuri82) [reply core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt]: yes, but that is the reason for why there is a `+ 10`, so a `0` on which `-1` is applied become a `9` :)

Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt]: I wrote the code to:
            "d" -> ((value.toInt() + randomness.choose(listOf(1,-1)) + 10) % 10).toString()

File: core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt (hunk 2 above)
Comment (jgaleotti) [reply core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt]: Indeed. DIsjunctionListRxGene.randomize() can leave un-initialised the inactive disjunction. Due to this, if standardMutation() changes the activeDisjunction, the gene can be left without randomising.

File: core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt [hunk 4]
```diff
@@ -44,9 +43,14 @@ class CharacterClassEscapeRxGene(
...
     override fun standardMutation(randomness: Randomness, apc: AdaptiveParameterControl, allGenes: List<Gene>) {
+        if (value=="") {
+            // if standardMutation was invoked before calling to randomize
+            // then we signal an exception
+            throw IllegalStateException("Cannot apply mutation on an uninitalized gene")
+        }
 
         value = when(type){
-            "d" -> ((value.toInt() + randomness.choose(listOf(1,-1) + 10)) % 10).toString()
+            "d" -> ((value.toInt() + randomness.choose(listOf(1,-1)) + 10) % 10).toString()
```
Comment (arcuri82) [inline core/src/main/kotlin/org/evomaster/core/search/gene/regex/CharacterClassEscapeRxGene.kt]: indeed, it was a bug! thx for fixing it ;)

**Pull request merged**